Use League\Flysystem\Exception in Adapters, users can catch single exception

mhlavac opened the pull request:
First of all, excellent work on the library, thanks everyone who is contributing.

The goal of this PR is to suggest using `League\Flysystem\Exception` for all errors coming from adapters. The change is related to issue #620 and #611 

**Why?**
I need to differentiate between false and exception. Also when I call `Filesystem->has()` or different method I want to put the logic inside try/catch block and act on it.

Right now different exceptions might be returned and there is no rule that custom adapters should extend or wrap exceptions inside `League\Flysystem\Exception`. This means that if I use `Filesystem` and I want to catch Google Cloud Storage exception I have to check Google SDK exceptions.

The fix for adapter can be implemented without BC break, I can extend or wrap exceptions in already existing `League\Flysystem\Exception`. So here I suggest changing the phpdocs to include `@throws Exception`.

**Possible BC break**
I would love to keep BC breaks minimal or none if possible, otherwise this will end up in 2.0 branch, but I changed Local and FTP adapters to throw `League\Flysystem\Exception`. I am happy to remove these changes from this PR if it's too much.

**Discussion**
Let's discuss what's the best approach to deal with this. Right now I think it's should be the Adapter's job to deal with specific exceptions and throwing something Filesystem user expects, but maybe there is better solution to this.

prisis commented:
Pls dont change the Runtime or Logic exceptions!
Create for all used base exception a flysystem exception, like League\Flysystem\RuntimeException or League\Flysystem\LogicException.

mhlavac commented:
@prisis thanks for the feedback. I introduced new Flysystem exceptions RuntimeException, LogicException and ErrorException.

There is no Exception interface at the moment, but they are all extending the `League\Flysystem\Exception` same as the other exceptions do.

I also updated the tests to reflect the new behavior.

prisis commented:
Much better :) 
But you need to change the throws on the interface with the right exception, Exception is only the base class :)

mhlavac commented:
@prisis sorry, this took me a while. Thanks for feedback again.

I updated the exception descriptions. For me it's important to tell apart if file doesn't exist or if it there was a connection error. Adapters can throw more specific exceptions, by extending the generic Exception.

Should I introduce `AdapterException` for that matter, or is using `Exception` good enough?
